talisker/context.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (C) 2016- Canonical Ltd
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+from builtins import *  # noqa
+
+try:
+    from collections.abc import Mapping
+except:  # < py3.3
+    from collections import Mapping
+
+from collections import OrderedDict
+from contextlib import contextmanager
+
+from werkzeug.local import Local, release_local
+from werkzeug.wsgi import ClosingIterator
+
+# a per request/job context. Generally, this will be the equivalent of thread
+# local storage, but if greenlets are being used it will be a greenlet local.
+context = Local()
+
+
+def clear():
+    release_local(context)
+
+
+def wsgi_middleware(app):
+    def middleware(environ, start_response):
+        return ClosingIterator(app(environ, start_response), clear)
+    return middleware
+
+
+class ContextStack(Mapping):
+    """A stacked set of dicts stored in a context.
+
+    Support lookups and iteration, which go from top of the stack down.
+    Can also be used as a context manager.
+
+    """
+    def __init__(self, name, *dicts):
+        """Initialise stack, with name to use in context storage."""
+        self.name = name
+        self._stack.extend(dicts)
+
+    def clear(self):
+        """Clear the stack."""
+        storage = {'stack': [], 'flattened': None}
+        setattr(context, self.name, storage)
+        return storage
```

Comment:
Do you really want to return the internal implementation's storage to the caller? Is `clear` useful anyway? If so, maybe this should be split into a public `clear` (that doesn't return the storage) and internal `_clear` that `clear` and `_storage` can call.